### PR TITLE
Add title and category editing to the GUI

### PR DESCRIPTION
**Pull request by ichorid**
This PR adds the ability to edit a GigaChannel entry through the GUI. The editing is done inline in the MyChannel table. The user can change the torrent's category by selecting one from a combo box. Torrent's title can also be edited by double-clicking on it.
Also, this PR changes Legacy entries status label from 6 to 1000, and removes FTS indexing from `tags` column, so the database format is not fully compatible with the previous versions.

This PR is really important for 7.3 release, as it provides a basis for crowdsourcing the torrents channels.

EDIT:
Also, this PR includes a refactoring of `triblertablecontrollers.py` and **breaking changes to DB schema** (removing unused tables and properties).

### CODE REVIEW

**Comment by devos50** [comment]
@ichorid note that the tests are failing and that there are several flake8 errors.

**Comment by ichorid** [comment]
> @ichorid note that the tests are failing and that there are several flake8 errors.

That's why it is still "draft" pull request :wink:

**Comment by devos50** [comment]
Ah, I got already requested to do a review so that’s why I asked.

**Comment by ichorid** [comment]
> Ah, I got already requested to do a review so that’s why I asked.

oh, sorry. Next time I will only do this when its really finished.

**Comment by devos50** [comment]
Well, it’s ok if you want a preliminary review round for instance (i.e. to give some high level feedback) ;)

**Comment by ichorid** [comment]
> Well, it’s ok if you want a preliminary review round for instance (i.e. to give some high level feedback) ;)

Yep, that was my intention here. I signalled you by requesting a review "*hey, look at this*", but there was a warning that this thing is still in the works (*draft* PR).

File: Tribler/Core/Category/category.conf [hunk 1]
```diff
@@ -12,7 +12,7 @@ matchpercentage = 0.5
 
 [VideoClips]
 rank = 2
-displayname = Video Clips
+displayname = VideoClips
```
Comment (devos50) [inline Tribler/Core/Category/category.conf]: Why not display it with the proper spelling (which is with a space between the words)?

File: Tribler/Core/Modules/MetadataStore/OrmBindings/channel_node.py [hunk 2]
```diff
@@ -19,7 +19,8 @@
 JUST_RECEIVED = 3
 UPDATE_AVAILABLE = 4
 PREVIEW_UPDATE_AVAILABLE = 5
-LEGACY_ENTRY = 6
+UPDATED = 6
```
Comment (devos50) [inline Tribler/Core/Modules/MetadataStore/OrmBindings/channel_node.py]: Could you please add some comments here to indicate what each status exactly means?

File: Tribler/Core/Modules/MetadataStore/store.py [hunk 3]
```diff
@@ -34,14 +33,13 @@
 # maintained by SQL triggers.
 sql_create_fts_table = """
     CREATE VIRTUAL TABLE IF NOT EXISTS FtsIndex USING FTS5
-        (title, tags, content='ChannelNode', prefix = '2 3 4 5',
+        (title, content='ChannelNode', prefix = '2 3 4 5',
```
Comment (devos50) [inline Tribler/Core/Modules/MetadataStore/store.py]: Do we have to remove our metadata database to run this?

File: Tribler/Core/Modules/restapi/metadata_endpoint.py [hunk 4]
```diff
@@ -101,7 +101,7 @@ def render_GET(self, request):
             channels_list = [channel.to_simple_dict() for channel in channels]
 
         return json.dumps({
-            "channels": channels_list,
+            "results": channels_list,
```
Comment (devos50) [inline Tribler/Core/Modules/restapi/metadata_endpoint.py]: I am not entirely sure whether I agree with this change. On one hand, it is more consistent. On the other hand, a `results` key is quite generic. What is the motivation behind this change?

File: Tribler/Core/Modules/restapi/mychannel_endpoint.py [hunk 5]
```diff
@@ -329,10 +327,14 @@ def __init__(self, session, infohash):
     @db_session
     def render_PATCH(self, request):
         parameters = http.parse_qs(request.content.read(), 1)
-        if 'status' not in parameters:
+        if 'status' not in parameters and 'tags' not in parameters and 'title' not in parameters:
```
Comment (devos50) [inline Tribler/Core/Modules/restapi/mychannel_endpoint.py]: The error text does not reflect the conditional (since it does not only check for status). Also, could you please document this method and give an example of a request? In particular, we should communicate the type of the `tags` parameter to the developer.

File: Tribler/Core/Modules/restapi/search_endpoint.py [hunk 6]
```diff
@@ -116,6 +116,8 @@ def search_db():
 
         def on_search_results(search_results_tuple):
             search_results, total = search_results_tuple
+            if not search_results:
```
Comment (devos50) [inline Tribler/Core/Modules/restapi/search_endpoint.py]: There's no JSON response here?

File: Tribler/Test/Core/Modules/MetadataStore/test_store.py [hunk 7]
```diff
@@ -229,6 +230,35 @@ def test_process_payload_reject_old(self):
...
+
+        torrent3 = self.mds.TorrentMetadata(title='blabla', timestamp=12, id_=4,
+                                            infohash=payload.infohash)
+        payload3 = torrent._payload_class(**torrent3.to_dict())
+        torrent3.delete()
+
+        self.mds.process_payload(payload2)
+        self.assertEqual(GOT_NEWER_VERSION, self.mds.process_payload(payload)[0][1])
+
+        # In this corner case the newle arrived payload contains a newer node
```
Comment (devos50) [inline Tribler/Test/Core/Modules/MetadataStore/test_store.py]: newle -> newly

File: Tribler/Test/Core/Modules/RestApi/test_mychannel_endpoint.py [hunk 8]
```diff
@@ -222,19 +220,19 @@ def test_update_my_torrents_no_channel(self):
         return self.do_request('mychannel/torrents', request_type='POST', post_data=post_params, expected_code=404)
 
     @trial_timeout(10)
-    def test_update_my_torrents(self):
+    def test_update_my_torrents_POST(self):
         """
-        Test whether we get an error if we update multiple torrents in your uncreated channel
+        Test updating/creating a torrent in a personal channel with POST
```
Comment (devos50) [inline Tribler/Test/Core/Modules/RestApi/test_mychannel_endpoint.py]: with POST -> with a POST request (minor comment)

File: TriblerGUI/qt_resources/mainwindow.ui [hunk 9]
```diff
@@ -3307,7 +3310,7 @@ font-weight: bold;</string>
               </widget>
              </item>
              <item>
-              <widget class="QLabel" name="num_search_results_label">
+              <widget class="QLabel" name="num_results_label">
```
Comment (devos50) [inline TriblerGUI/qt_resources/mainwindow.ui]: In Qt, names of `QWidget` objects should be unique across the whole `.ui` file. That's why I name these objects as specific as possible. The name `num_results_label` could result in a name collision.

File: TriblerGUI/widgets/channelpage.py [hunk 10]
```diff
@@ -50,4 +52,4 @@ def initialize_with_channel(self, channel_info):
 
     def load_torrents(self):
         self.controller.model.reset()
-        self.controller.load_torrents(1, 50)  # Load the first 50 torrents
+        self.controller.perform_query(first=1, last=50)  # Load the first 50 torrents
```
Comment (devos50) [inline TriblerGUI/widgets/channelpage.py]: I like using the full keyword argument name here 👍

File: TriblerGUI/widgets/discoveredpage.py [hunk 11]
```diff
@@ -36,4 +36,4 @@ def initialize_discovered_page(self, gui_settings):
 
     def load_discovered_channels(self):
         self.controller.model.reset()
-        self.controller.load_channels(1, 50)  # Load the first 50 discovered channels
+        self.controller.perform_query(start=1, end=50)  # Load the first 50 discovered channels
```
Comment (devos50) [inline TriblerGUI/widgets/discoveredpage.py]: Probably not related to changes in this PR, but we are using start and end here as variable names and in `channelpage.py` we are using first and last. We should make these names consistent.

File: TriblerGUI/widgets/lazytableview.py [hunk 12]
```diff
@@ -82,7 +82,7 @@ def on_subscribe_control_clicked(self, index):
         item = index.model().data_items[index.row()]
         # skip LEGACY entries, regular torrents and personal channel
         if (u'subscribed' not in item or
-                item[u'status'] == 6 or
+                item[u'status'] == 1000 or
```
Comment (devos50) [inline TriblerGUI/widgets/lazytableview.py]: We should re-use the metadata state definitions here.

File: TriblerGUI/widgets/tablecontentdelegate.py [hunk 13]
```diff
@@ -4,13 +4,29 @@
...
-from PyQt5.QtWidgets import QStyle, QStyledItemDelegate
+from PyQt5.QtWidgets import QStyle, QStyledItemDelegate, QComboBox
 
 from TriblerGUI.defs import ACTION_BUTTONS, COMMIT_STATUS_COMMITTED, COMMIT_STATUS_NEW, COMMIT_STATUS_TODELETE, \
     HEALTH_CHECKING, HEALTH_DEAD, HEALTH_ERROR, HEALTH_GOOD, HEALTH_MOOT, HEALTH_UNCHECKED
 from TriblerGUI.utilities import get_health, get_image_path
 from TriblerGUI.widgets.tableiconbuttons import DeleteIconButton, DownloadIconButton, PlayIconButton
 
+#TODO: build this automatically and/or move it somewhere
+CATEGORY_LIST = [
```
Comment (devos50) [inline TriblerGUI/widgets/tablecontentdelegate.py]: I would move these definitions to the `defs.py` file in the `TriblerGUI` package.

File: Tribler/Core/Category/category.conf (hunk 1 above)
Comment (ichorid) [reply Tribler/Core/Category/category.conf]: This thing is only used to tag Legacy channels now. However, the tag should be a single word. Hence, this change.

File: Tribler/Core/Modules/MetadataStore/OrmBindings/channel_node.py (hunk 2 above)
Comment (ichorid) [reply Tribler/Core/Modules/MetadataStore/OrmBindings/channel_node.py]: Good point, I'll do that.

File: Tribler/Core/Modules/MetadataStore/store.py (hunk 3 above)
Comment (ichorid) [reply Tribler/Core/Modules/MetadataStore/store.py]: Actually, we don't have to. But it is a very good idea to do so because your FTS index gets much more balanced afterward.

File: TriblerGUI/widgets/tablecontentdelegate.py (hunk 13 above)
Comment (ichorid) [reply TriblerGUI/widgets/tablecontentdelegate.py]: Good point!

File: TriblerGUI/widgets/lazytableview.py (hunk 12 above)
Comment (ichorid) [reply TriblerGUI/widgets/lazytableview.py]: But I guess you were against adding dependencies to TrilberGUI from the Core, right?

File: Tribler/Core/Modules/restapi/metadata_endpoint.py (hunk 4 above)
Comment (ichorid) [reply Tribler/Core/Modules/restapi/metadata_endpoint.py]: The reason is the big refactoring of `triblertablecontrollers.py`. Having the API consistent across different endpoints really helps with inheritance and code reuse. Both for API writers and users.

File: TriblerGUI/qt_resources/mainwindow.ui (hunk 9 above)
Comment (ichorid) [reply TriblerGUI/qt_resources/mainwindow.ui]: Hmm. There are not collisions now. I guess we're fine with this name now. We can always rename it later.

File: TriblerGUI/widgets/discoveredpage.py (hunk 11 above)
Comment (ichorid) [reply TriblerGUI/widgets/discoveredpage.py]: Agree.

File: Tribler/Core/Modules/restapi/search_endpoint.py (hunk 6 above)
Comment (ichorid) [reply Tribler/Core/Modules/restapi/search_endpoint.py]: Hmm... Maybe you're right, and there should be a zero-results response instead. I used this thing as a temporary workaround for a problem.

File: Tribler/Core/Modules/restapi/mychannel_endpoint.py (hunk 5 above)
Comment (ichorid) [reply Tribler/Core/Modules/restapi/mychannel_endpoint.py]: Done;

File: Tribler/Core/Modules/MetadataStore/store.py (hunk 3 above)
Comment (ichorid) [reply Tribler/Core/Modules/MetadataStore/store.py]: Ok, now I've broken the schema. So, we have to remove our database.

File: TriblerGUI/widgets/lazytableview.py (hunk 12 above)
Comment (ichorid) [reply TriblerGUI/widgets/lazytableview.py]: Hm. I'll just add them to `defs.py`.

Comment (devos50) [reply TriblerGUI/widgets/lazytableview.py]: Yes but at this point I rather have a link to the core (which we can later refactor) instead of having a magic variable which is unclear.

**Comment by devos50** [comment]
@ichorid one more thing, since you are changing the REST API response structure, please also make the appropriate changes to the fake Tribler API (https://github.com/Tribler/fake-tribler-api), in particular [here](https://github.com/Tribler/fake-tribler-api/blob/devel/FakeTriblerAPI/endpoints/metadata_endpoint.py#L78) and [here](https://github.com/Tribler/fake-tribler-api/blob/devel/FakeTriblerAPI/endpoints/metadata_endpoint.py#L153).

**Comment by devos50** [comment]
@ichorid additional comment, please do not try to squeeze too much in PRs. For now you can continue but enabling changing the category and torrent name could be in one PR, and the refactoring of the Tribler table controllers in the GUI could have been in another PR 👍

**Comment by ichorid** [comment]
> @ichorid additional comment, please do not try to squeeze too much in PRs. For now you can continue but enabling changing the category and torrent name could be in one PR, and the refactoring of the Tribler table controllers in the GUI could have been in another PR +1

Yeah, I was thinking about this too... However, these things depend on each other, so I really can't pipeline them. And I'm running out of time before my trip.
Also, would you really prefer 3 different PRs that break DB schema instead of one :wink:

**Comment by ichorid** [comment]
Corresponding Fake Tribler API PR: https://github.com/Tribler/fake-tribler-api/pull/47

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
retest this please
